### PR TITLE
[Heartbeat] Dedupe screenshots / Extra Args

### PR DESCRIPTION
**Pull request by andrewvc**
Counterpart of https://github.com/elastic/synthetics/pull/282
Fixes: https://github.com/elastic/beats/issues/26188

Adds support for screenshot blocks as mentioned in the synthetics PR, and also switches over to using the new flags / capabilities system. This is done by using `--rich-events` by default as a synthetics flag, and also exposing a new `synthetics_args` option for synthetics, allowing users to use configs like:

```yaml
- type: browser
  enabled: true
  id: my-monitor
  name: My Monitor
  synthetics_args: ["--capability", "trace", "metrics", "filmstrips", "ssblocks"]
  source:
    inline:
      script:
        step("load homepage", async () => {
            await page.goto('https://www.elastic.co');
        });
        step("hover over products menu", async () => {
            await page.hover('css=[data-nav-item=products]');
        });
        step("failme", async () => {
            await page.hhover('css=[data-nav-item=products]');
        });
  schedule: "@every 1m"
```

This PR also moves some of the logic from `browser.go` to `suite.go`, which was also renamed to be cleaner and more in keeping with idiomatic go. This also made adding a small test possible. We still need a better story around E2E testing, but for now this will suffice.

Finally, this adds mappings for `screenshot_ref` objects as well as blocks, and routes them to the correct dataset. It also improves tests for enrichment of events to help test this.

### CODE REVIEW

**Review by vigneshshanmugam — COMMENTED** [review]
LGTM, tested and it works as expected. 

We might need to change one more thing to get it merged. adding `step/block` to the screenshot datastream index.

**Comment by andrewvc** [comment]
PR now adds mapping and properly routes to correct dataset. Still needs tests

File: heartbeat/_meta/fields.common.yml [hunk 1]
```diff
@@ -174,6 +174,22 @@
...
+            - name: width
+              type: integer
+              description: Width of the full screenshot in pixels.
+            - name: height
+              type: integer
+              description: Height of the full screenshot in pixels
+            - name: blocks
+              type: group
+              fields: 
+                - name: hash
```
Comment (vigneshshanmugam) [inline heartbeat/_meta/fields.common.yml:190]: Need to add `top` and `left` here.

File: heartbeat/_meta/fields.common.yml [hunk 2]
```diff
@@ -174,6 +174,22 @@
               type: text
             - name: stack
               type: text
+        - name: screenshot_ref
+          type: group
+          dynamic: false
+          fields:
+            - name: width
+              type: integer
+              description: Width of the full screenshot in pixels.
```
Comment (vigneshshanmugam) [inline heartbeat/_meta/fields.common.yml:183]: Shall we remove the use of full here? Full seems like full page screenshot but we are capture only the viewport width and height.

File: x-pack/heartbeat/monitors/browser/synthexec/synthexec.go [hunk 3]
```diff
@@ -108,23 +109,31 @@ func runCmd(
 		return nil, err
 	}
 
+	// Default capabilities
+	capabilities = append(capabilities, "trace", "metrics", "filmstrips", "ssblocks")
```
Comment (vigneshshanmugam) [inline x-pack/heartbeat/monitors/browser/synthexec/synthexec.go:113]: I am not sure if its the right way to do it. 

I was thinking we should make it configurable by introducing Capability flag in the `config.go` in addition to params and then using it here for experimentation. 

moving forward once the feature is built in to the UI, we can add it to `--rich-events` in synthetics. 

WDYT?

File: x-pack/heartbeat/monitors/browser/synthexec/synthexec.go [hunk 4]
```diff
@@ -168,9 +177,6 @@ func runCmd(
 	// Close mpx after the process is done and all events have been sent / consumed
 	go func() {
 		err := cmd.Wait()
-		if err != nil {
```
Comment (vigneshshanmugam) [inline x-pack/heartbeat/monitors/browser/synthexec/synthexec.go:171]: Do we do anything special now for handling errors messages?

File: heartbeat/_meta/fields.common.yml (hunk 2 above)
Comment (andrewvc) [reply heartbeat/_meta/fields.common.yml:183]: Hmmm, I intended it to mean not of an individual block. I'll clarify that.

File: heartbeat/_meta/fields.common.yml (hunk 1 above)
Comment (andrewvc) [reply heartbeat/_meta/fields.common.yml:190]: Good catch!

File: x-pack/heartbeat/monitors/browser/synthexec/synthexec.go (hunk 3 above)
Comment (andrewvc) [reply x-pack/heartbeat/monitors/browser/synthexec/synthexec.go:113]: Ah, I see what you mean. That makes sense, will make those changes!

File: x-pack/heartbeat/monitors/browser/synthexec/synthexec.go (hunk 4 above)
Comment (vigneshshanmugam) [reply x-pack/heartbeat/monitors/browser/synthexec/synthexec.go:171]: Sure, thanks for the info.

File: x-pack/heartbeat/monitors/browser/suite.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package browser
```
Comment (andrewvc) [inline x-pack/heartbeat/monitors/browser/suite.go:5]: This file isn't new, it's just moved. Enough of it changed due to renaming the `ss` var to `s` that git thinks it's all new.

File: CHANGELOG.next.asciidoc [hunk 6]
```diff
@@ -106,6 +106,7 @@ https://github.com/elastic/beats/compare/v7.0.0-alpha2...master[Check the HEAD d
 - Change logging in logs input to structure logging. Some log message formats have changed. {pull}25299[25299]
 
 *Heartbeat*
+- Add support for screenshot blocks and use newer synthetics flags that only worker in newer synthetics betas. {pull}25808[25808]
```
Comment (vigneshshanmugam) [inline CHANGELOG.next.asciidoc:109]:
```suggestion
- Add support for screenshot blocks and use newer synthetics flags that only works in newer synthetics betas. {pull}25808[25808]
```

File: heartbeat/_meta/fields.common.yml (hunk 1 above)
Comment (vigneshshanmugam) [reply heartbeat/_meta/fields.common.yml:190]: I don't see the `top` & `left` yet in the new changes.

File: x-pack/heartbeat/monitors/browser/suite_test.go [hunk 7]
```diff
@@ -114,3 +115,37 @@ func TestEmptySource(t *testing.T) {
...
+			nil,
+		},
+		{
+			"sandbox",
+			&Config{Sandbox: true},
+			[]string{"--sandbox"},
+		},
+		{
+			"kitchen sink",
+			&Config{SyntheticsArgs: []string{"--capability", "trace"}, Sandbox: true},
```
Comment (vigneshshanmugam) [inline x-pack/heartbeat/monitors/browser/suite_test.go:137]: I am curios to see a test with capability > 2 and other flags as part of synthetic args. Reason being variadic args are prone to errors.

**Review by vigneshshanmugam — APPROVED** [review]
LGTM after the `top` and `left` changes.

**Comment by andrewvc** [comment]
@vigneshshanmugam I believe I've addressed your concerns. I left out `top` and `left` on the mapping intentionally. We'll probably never query them, so it makes sense to save a field. I've added some notes to the fields yml indicating why this is.

**Comment by vigneshshanmugam** [comment]
@andrewvc Spot on, Sounds good to me ++.

**Comment by andrewvc** [comment]
@Mergifyio backport 7.x

**Comment by andrewvc** [comment]
@Mergifyio backport 7.14

**Pull request merged**